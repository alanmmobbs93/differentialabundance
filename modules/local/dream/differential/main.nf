// TODO:
//      should this module take all variables already processed (meta),
//      or should it be able to take the yaml file and parse it on its own?

process DREAM_DIFFERENTIAL {
    tag "${meta.id}"
    label 'process_single'

    conda "${moduleDir}/environment.yml" // TODO: COMPLETE THIS TOO!
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/20/201c970c051ff973190938c6fb75ce99530346115e8a7600b9423be00daf3fdf/data' :
        'community.wave.seqera.io/library/bioconductor-edger_bioconductor-variancepartition:69f4dda505cec1f3' }"

    input:
    val meta // array with id, [...]
    tuple val(meta2), path(samplesheet), path(intensities)

    output: // TODO: COMPLETE OUTPUTS!
    tuple val(meta), path("*.limma.results.tsv")          , emit: results
    tuple val(meta), path("*.limma.mean_difference.png")  , emit: md_plot
    tuple val(meta), path("*.MArrayLM.limma.rds")         , emit: rdata
    tuple val(meta), path("*.limma.model.txt")            , emit: model
    tuple val(meta), path("*.R_sessionInfo.log")          , emit: session_info
    tuple val(meta), path("*.normalised_counts.tsv")      , emit: normalised_counts, optional: true
    path "versions.yml"                                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix                 = task.ext.prefix ?: meta.id
    def blocking_factors       = meta.blocking_factors ? "--blocking_variables ${meta.blocking_factors}" : ''
    def exclude_samples_col    = meta.exclude_samples_col ? "--exclude_samples_col ${meta.exclude_samples_col}" : ''
    def exclude_samples_values = meta.exclude_samples_values ? "--exclude_samples_values ${meta.exclude_samples_values}" : ''

    """
    ## TODO: CHECK WHERE DO exclude_samples_col exclude_samples_values and number OPTIONS COME FROM!!
    dream_de.R  \\
        --output_prefix ${prefix} \\
        --count_file ${intensities} \\
        --sample_file ${samplesheet} \\
        --contrast_variable ${meta.contrast_variable} \\
        --reference_level ${meta.reference} \\
        --target_level ${meta.target} \\
        ${blocking_factors} \\
        ${exclude_samples_col} \\
        ${exclude_samples_values} \\
        --threads ${task.cpus} \\
        --number 100 ## ??

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        r-optparse: \$(Rscript -e "library(optparse); cat(as.character(packageVersion('optparse')))")
        r-edgeR: \$(edgeR -e "library(edgeR); cat(as.character(packageVersion('edgeR')))")
        r-variancePartition: \$(Rscript -e "library(variancePartition); cat(as.character(packageVersion('variancePartition')))")
        r-BiocParallel: \$(Rscript -e "library(BiocParallel); cat(as.character(packageVersion('BiocParallel')))")
    END_VERSIONS
    """


    stub: // TODO: COMPLETE STUB!
    prefix              = task.ext.prefix   ?: "${meta.id}"
    """
    #!/usr/bin/env Rscript
    library(limma)
    a <- file("${prefix}.limma.results.tsv", "w")
    close(a)
    a <- file("${prefix}.limma.mean_difference.png", "w")
    close(a)
    a <- file("${prefix}.MArrayLM.limma.rds", "w")
    close(a)
    a <- file("${prefix}.normalised_counts.tsv", "w")
    close(a)
    a <- file("${prefix}.limma.model.txt", "w")
    close(a)
    a <- file("${prefix}.R_sessionInfo.log", "w")
    close(a)

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        r-optparse: \$(Rscript -e "library(optparse); cat(as.character(packageVersion('optparse')))")
        r-edgeR: \$(edgeR -e "library(edgeR); cat(as.character(packageVersion('edgeR')))")
        r-variancePartition: \$(Rscript -e "library(variancePartition); cat(as.character(packageVersion('variancePartition')))")
        r-BiocParallel: \$(Rscript -e "library(BiocParallel); cat(as.character(packageVersion('BiocParallel')))")
    END_VERSIONS
    """
}
