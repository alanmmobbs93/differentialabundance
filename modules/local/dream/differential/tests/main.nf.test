nextflow_process {

    name "Test Process DREAM_DIFFERENTIAL"
    script "../main.nf"
    process "DREAM_DIFFERENTIAL"
    config "./nextflow.config"

    tag "modules"
    tag "models"
    tag "dge"
    tag "dream"
    tag "dream/differential"

    test("Mus musculus - expression table - contrasts") {

        when {
            process {
                """
                // Metadata
                input[0] = Channel.of(
                        [
                            id: "treatment_mCherry_hND6",
                            contrast_variable: "treatment",
                            reference: "mCherry",
                            target: "hND6",
                            blocking_factors: null
                        ]
                )

                // expression data [ [meta], file(samplesheet), file(expression) ]
                input[1] = Channel.of(
                    [
                        [id: 'test' ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Mus musculus - expression table - contrasts + blocking factors") {

        when {
            process {
                """
                // Metadata
                input[0] = Channel.of(
                        [
                            id: "treatment_mCherry_hND6",
                            contrast_variable: "treatment",
                            reference: "mCherry",
                            target: "hND6",
                            blocking_factors: 'sample_number'
                        ]
                )

                // expression data [ [meta], file(samplesheet), file(expression) ]
                input[1] = Channel.of(
                    [
                        [id: 'test' ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Mus musculus - expression table - contrasts + blocking factors stub") {

        options '-stub'
        when {
            process {
                """
                // Metadata
                input[0] = Channel.of(
                        [
                            id: "treatment_mCherry_hND6",
                            contrast_variable: "treatment",
                            reference: "mCherry",
                            target: "hND6",
                            blocking_factors: 'sample_number'
                        ]
                )

                // expression data [ [meta], file(samplesheet), file(expression) ]
                input[1] = Channel.of(
                    [
                        [id: 'test' ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
