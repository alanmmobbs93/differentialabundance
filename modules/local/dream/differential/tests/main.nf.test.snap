{
    "Mus musculus - expression table - contrasts + blocking factors": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,bd28cb586bb40ad9f0b7069cb3ada0a2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,0d1c235cca14dcf75af4ef6cb66247ae"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,9197daf020341f1fc92636c762cb40e0"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,1fab14c3a7fb300bc6d8e693ef75cf27"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,e591353ef3f5d63a74889d57f7ce37c6"
                    ]
                ],
                "6": [
                    
                ],
                "7": [
                    "versions.yml:md5,56f3e82916abb64aa39ae9d008d71978"
                ],
                "contrasts_png": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,1fab14c3a7fb300bc6d8e693ef75cf27"
                    ]
                ],
                "md_plot": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,0d1c235cca14dcf75af4ef6cb66247ae"
                    ]
                ],
                "model": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                    ]
                ],
                "normalised_counts": [
                    
                ],
                "rdata": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,9197daf020341f1fc92636c762cb40e0"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,bd28cb586bb40ad9f0b7069cb3ada0a2"
                    ]
                ],
                "session_info": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,e591353ef3f5d63a74889d57f7ce37c6"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,56f3e82916abb64aa39ae9d008d71978"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-27T18:45:31.211486573"
    },
    "Mus musculus - expression table - contrasts + blocking factors stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    
                ],
                "7": [
                    "versions.yml:md5,8338b523d0f508917db713c8a0cb7fdf"
                ],
                "contrasts_png": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "md_plot": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "model": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "normalised_counts": [
                    
                ],
                "rdata": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "session_info": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8338b523d0f508917db713c8a0cb7fdf"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-27T18:46:04.900211679"
    },
    "Mus musculus - expression table - contrasts": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,bd28cb586bb40ad9f0b7069cb3ada0a2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,0d1c235cca14dcf75af4ef6cb66247ae"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,9197daf020341f1fc92636c762cb40e0"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,1fab14c3a7fb300bc6d8e693ef75cf27"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,e591353ef3f5d63a74889d57f7ce37c6"
                    ]
                ],
                "6": [
                    
                ],
                "7": [
                    "versions.yml:md5,56f3e82916abb64aa39ae9d008d71978"
                ],
                "contrasts_png": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.contrasts_plot.png:md5,1fab14c3a7fb300bc6d8e693ef75cf27"
                    ]
                ],
                "md_plot": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.mean_difference.png:md5,0d1c235cca14dcf75af4ef6cb66247ae"
                    ]
                ],
                "model": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.model.txt:md5,d2113d82b76046c319e6602da2ad74d6"
                    ]
                ],
                "normalised_counts": [
                    
                ],
                "rdata": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.MArrayMM.dream.rds:md5,9197daf020341f1fc92636c762cb40e0"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,bd28cb586bb40ad9f0b7069cb3ada0a2"
                    ]
                ],
                "session_info": [
                    [
                        {
                            "id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "reference": "mCherry",
                            "target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.R_sessionInfo.log:md5,e591353ef3f5d63a74889d57f7ce37c6"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,56f3e82916abb64aa39ae9d008d71978"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-27T18:44:52.310595802"
    }
}